xds/src/main/java/io/grpc/xds/GoogleCloudToProdNameResolver.java

```diff
@@ -0,0 +1,260 @@
+      return;
+    }
+    resolving = true;
+    if (!schemeOverride.equals("xds")) {
+      delegate.start(listener);
+      succeeded = true;
+      resolving = false;
+      return;
+    }
+    if (executor == null) {
+      executor = SharedResourceHolder.get(executorResource);
+    }
+
+    class Resolve implements Runnable {
+      @Override
+      public void run() {
+        String zone;
+        boolean supportIpv6;
+        ImmutableMap<String, ?> rawBootstrap = null;
+        try {
+          zone = queryZoneMetadata(METADATA_URL_ZONE);
+          supportIpv6 = queryIpv6SupportMetadata(METADATA_URL_SUPPORT_IPV6);
+          rawBootstrap = generateBootstrap(zone, supportIpv6);
+        } catch (IOException e) {
+          listener.onError(Status.INTERNAL.withDescription("Unable to get metadata").withCause(e));
+        } finally {
+          final ImmutableMap<String, ?> finalRawBootstrap = rawBootstrap;
+          syncContext.execute(new Runnable() {
+            @Override
+            public void run() {
+              if (!shutdown && finalRawBootstrap != null) {
+                xdsClientPoolFactory.setBootstrapOverride(finalRawBootstrap);
+                delegate.start(listener);
+                succeeded = true;
+              }
+              resolving = false;
+            }
+          });
+        }
+      }
+    }
+
+    executor.execute(new Resolve());
+  }
+
+  private static ImmutableMap<String, ?> generateBootstrap(String zone, boolean supportIpv6) {
+    ImmutableMap.Builder<String, Object> nodeBuilder = ImmutableMap.builder();
+    nodeBuilder.put("id", "C2P");
+    if (!zone.isEmpty()) {
+      nodeBuilder.put("locality", ImmutableMap.of("zone", zone));
+    }
+    if (supportIpv6) {
+      nodeBuilder.put("metadata",
+          ImmutableMap.of("TRAFFICDIRECTOR_DIRECTPATH_C2P_IPV6_CAPABLE", true));
+    }
+    ImmutableMap.Builder<String, Object> serverBuilder = ImmutableMap.builder();
+    serverBuilder.put("server_uri", "directpath-trafficdirector.googleapis.com");
+    serverBuilder.put("channel_creds",
+        ImmutableList.of(ImmutableMap.of("type", "google_default")));
+    return ImmutableMap.of(
+        "node", nodeBuilder.build(),
+        "xds_servers", ImmutableList.of(serverBuilder.build()));
+  }
+
+  @Override
+  public void refresh() {
+    if (!resolving && delegate != null) {
+      if (succeeded) {
+        delegate.refresh();
+      } else {
+        resolve();
+      }
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    if (shutdown) {
+      return;
+    }
+    shutdown = true;
+    if (delegate != null) {
+      delegate.shutdown();
+    }
+    if (executor != null && usingExecutorResource) {
+      executor = SharedResourceHolder.release(executorResource, executor);
+    }
+  }
+
+  private String queryZoneMetadata(String url) throws IOException {
+    HttpURLConnection con = httpConnectionProvider.createConnection(url);
+    if (con.getResponseCode() != 200) {
```

Comment:
Could response code be 202 and eventually a response body is returned?

Reply:
I thought about it as well. But I am just following https://github.com/grpc/grpc/blob/3944bfaf76b3123d40c68d29e3c1eeeed08afe57/src/core/ext/filters/client_channel/resolver/google_c2p/google_c2p_resolver.cc#L220